cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java

```diff
@@ -0,0 +1,415 @@
+  }
+
+  /**
+   * Creates a decoder for the key type.
+   *
+   * @param type type to decode to
+   */
+  private Function<ByteBuffer, KEY> createKeyDecoder(Type type) {
+    return createDecoder(type, "No decoder for decoding message key");
+  }
+
+  /**
+   * Creates a decoder for the payload type.
+   *
+   * @param type type to decode to
+   */
+  private Function<ByteBuffer, PAYLOAD> createPayloadDecoder(Type type) {
+    return createDecoder(type, "No decoder for decoding message payload");
+  }
+
+  /**
+   * Creates a decoder for decoding {@link ByteBuffer} for known type. It supports
+   * <p/>
+   * <pre>
+   * - String (assuming UTF-8)
+   * - byte[]
+   * - ByteBuffer
+   * </pre>
+   *
+   * @param type type to decode to
+   * @param failureDecodeMessage message for the exception if decoding of the given type is not supported
+   * @param <T> Type of the decoded type
+   * @return A {@link Function} that decode {@link ByteBuffer} into the given type or a failure decoder created through
+   *         {@link #createFailureDecoder(String)} if the type is not support
+   */
+  @SuppressWarnings("unchecked")
+  private <T> Function<ByteBuffer, T> createDecoder(Type type, String failureDecodeMessage) {
+    if (String.class.equals(type)) {
+      return (Function<ByteBuffer, T>) createStringDecoder();
+    }
+    if (ByteBuffer.class.equals(type)) {
+      return (Function<ByteBuffer, T>) createByteBufferDecoder();
+    }
+    if (byte[].class.equals(type) || (type instanceof GenericArrayType &&
+      byte.class.equals(((GenericArrayType) type).getGenericComponentType()))) {
+      return (Function<ByteBuffer, T>) createBytesDecoder();
+    }
+    return createFailureDecoder(failureDecodeMessage);
+  }
+
+  /**
+   * Creates a decoder that convert the input {@link ByteBuffer} into UTF-8 String. The input {@link ByteBuffer}
+   * will not be consumed after the call.
+   */
+  private Function<ByteBuffer, String> createStringDecoder() {
+    return new Function<ByteBuffer, String>() {
+      @Override
+      public String apply(ByteBuffer input) {
+        input.mark();
+        String result = Charsets.UTF_8.decode(input).toString();
+        input.reset();
+        return result;
+      }
+    };
+  }
+
+  /**
+   * Creates a decoder that returns the same input {@link ByteBuffer}.
+   */
+  private Function<ByteBuffer, ByteBuffer> createByteBufferDecoder() {
+    return new Function<ByteBuffer, ByteBuffer>() {
+      @Override
+      public ByteBuffer apply(ByteBuffer input) {
+        return input;
+      }
+    };
+  }
+
+  /**
+   * Creates a decoder that reads {@link ByteBuffer} content and return it as {@code byte[]}. The
+   * input {@link ByteBuffer} will not be consumed after the call.
+   */
+  private Function<ByteBuffer, byte[]> createBytesDecoder() {
+    return new Function<ByteBuffer, byte[]>() {
+      @Override
+      public byte[] apply(ByteBuffer input) {
+        byte[] bytes = new byte[input.remaining()];
```

Comment:
I think this is a bit different bc we are trying to read the cur position to end of buffer. Will keep it as is for now.